diskquota.c

```diff
@@ -638,53 +623,66 @@ disk_quota_launcher_main(Datum main_arg)
 	loop_end = time(NULL);
 
 	struct timeval nap;
-	nap.tv_sec                  = diskquota_naptime;
-	nap.tv_usec                 = 0;
-	TimestampTz loop_start_time = GetCurrentTimestamp();
+	nap.tv_sec  = diskquota_naptime;
+	nap.tv_usec = 0;
 	/* main loop: do this until the SIGTERM handler tells us to terminate. */
 	ereport(LOG, (errmsg("[diskquota launcher] start main loop")));
-	curDB = NULL;
+	DiskquotaDBEntry *curDB          = NULL;
+	Oid               curDBId        = 0;
+	bool              advance_one_db = true;
+	bool              timeout        = false;
 	while (!got_sigterm)
 	{
 		int rc;
 		CHECK_FOR_INTERRUPTS();
-
-		/*
-		 * modify wait time
-		 */
-		long secs;
-		int  microsecs;
-		TimestampDifference(GetCurrentTimestamp(),
-		                    TimestampTzPlusMilliseconds(loop_start_time, diskquota_naptime * 1000L), &secs, &microsecs);
-		nap.tv_sec  = secs;
-		nap.tv_usec = microsecs;
-
-		if (curDB == DiskquotaLauncherShmem->dbArrayTail)
+		/* pick a db to run */
+		if (advance_one_db)
 		{
-			/* Have sleep enough time, should start another loop */
-			if (nap.tv_sec == 0 && nap.tv_usec == 0)
+			curDB   = next_db(curDB);
+			timeout = false;
+			if (curDB != NULL)
 			{
-				loop_start_time = GetCurrentTimestamp();
-				/* set the curDB pointing to the head of the db list */
-				curDB = NULL;
+				curDBId = curDB->dbid;
+				elog(DEBUG1, "[diskquota] next db to run:%d", curDB->id);
 			}
-			/* do nothing, just to sleep untill the nap time is 0 */
 			else
-			{
-				continue;
-			}
+				elog(DEBUG1, "[diskquota] no db to run");
 		}
-
-		/* If there are no enough workers to run db, we can firstly sleep to wait workers */
-		if (nap.tv_sec == 0 && nap.tv_usec == 0)
+		/*
+		 * Modify wait time
+		 *
+		 * If there is no db needed to run or has exceeded the next_run_time,
+		 * just sleep to wait a db or a free worker.
+		 *
+		 * Otherwise check the next_run_time to determin how much time to wait
+		 */
+		if (timeout || curDB == NULL)
 		{
 			nap.tv_sec  = diskquota_naptime > 0 ? diskquota_naptime : 1;
 			nap.tv_usec = 0;
 		}
-
-		while (curDB != DiskquotaLauncherShmem->dbArrayTail && CanLaunchWorker())
+		else
 		{
-			start_worker();
+			long        secs;
+			int         microsecs;
+			TimestampTz curTime = GetCurrentTimestamp();
+			TimestampDifference(curTime, curDB->next_run_time, &secs, &microsecs);
+			nap.tv_sec  = secs;
+			nap.tv_usec = microsecs;
+
+			/* if the sleep time is too short, just skip the sleeping */
+			if (secs == 0 && microsecs < MIN_SLEEPTIME * 1000L)
+			{
+				nap.tv_usec = 0;
+			}
+
+			/* if the sleep time is too long, advance the next_run_time */
+			if (secs > diskquota_naptime)
+			{
+				nap.tv_sec           = diskquota_naptime;
+				nap.tv_usec          = 0;
+				curDB->next_run_time = TimestampTzPlusMilliseconds(curTime, diskquota_naptime * 1000L);
+			}
```

Comment:
```suggestion
			/* if the sleep time is too short, just skip the sleeping */
			if (secs == 0 && microsecs < MIN_SLEEPTIME * 1000L)
			{
				nap.tv_usec = 1;
				nap.tv_usec = 0;
				curDB->next_run_time = TimestampTzPlusMilliseconds(curTime, 1000L);
			}

			/* if the sleep time is too long, advance the next_run_time */
			if (nap.tv_secs >= diskquota_naptime)
			{
				nap.tv_sec           = diskquota_naptime;
				nap.tv_usec          = 0;
				curDB->next_run_time = TimestampTzPlusMilliseconds(curTime, diskquota_naptime * 1000L);
			}
```
The logic of this code is a little messy. Firstly, if the sleep time is too short, we should also set `nap.tv_usec` to 1, as the comment `wait at least one time slice, avoid 100% CPU usage` is written in the 705th line. Secondly, I think it is better to describe the compare statement as `nap.tv_sec > diskquota_naptime` in the second `if` statement.

Reply:
Then, we can remove `if (!diskquota_naptime) usleep(1);` in the 706th line.

Reply:
Hmm... this is a little complicated issue.

1. I will remove line 705th, as have set the nap and it's not 0, no need to check the diskquota.naptime anymore.
2. > Firstly, if the sleep time is too short, we should also set nap.tv_usec to 1, as the comment wait at least one time slice, avoid 100% CPU usage

I think when the diskquota launcher has nothing to do (no db or no free worker),  and the diskquota.nap time is 0, we need to let the diskquota launcher sleep for a while instead of polling and causing CPU busy. In other cases, there is no need to check if the diskquota.naptime is 0. ( Although there are a lot of checks  in diskquota now, but doesn't make sense). Such as if we set the diskquota.naptime to 0,  we should let it work without a break when there are dbs and enough free workers to run.